core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -1818,8 +1835,182 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+        case None => throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadGateway, s"Failed to retrieve Google Project Number for Google Project ${googleProjectId}"))
+        case Some(longProjectNumber) => GoogleProjectNumber(longProjectNumber.toString)
+      }
+    }
+  }
+
+  /**
+    * If a ServicePerimeter is specified on the BillingProject, then we should update the list of Google Projects in the
+    * Service Perimeter.  All newly created Workspaces (and their newly claimed Google Projects) should already be
+    * persisted in the Rawls database prior to calling this method.  If no ServicePerimeter is specified on the Billing
+    * Project, do nothing
+    *
+    * @param billingProject
+    * @param span
+    * @return Future[Unit]
+    */
+  private def maybeUpdateGoogleProjectsInPerimeter(billingProject: RawlsBillingProject, span: Span = null): Future[Unit] = {
+    billingProject.servicePerimeter match {
+      case Some(servicePerimeterName) => overwriteGoogleProjectsInPerimeter(servicePerimeterName)
+      case None => Future.successful()
+    }
+  }
+
+  /**
+    * Some Service Perimeters may required that they have some additional non-Terra Google Projects that need to be in
+    * the perimeter for some other reason.  These are provided to us by the Service Perimeter stakeholders and we add
+    * them to the Rawls Config so that whenever we update the list of projects for a perimeter, these projects are
+    * always included.
+    *
+    * @param servicePerimeterName
+    * @return
+    */
+  private def loadStaticProjectsForPerimeter(servicePerimeterName: ServicePerimeterName): Seq[GoogleProjectNumber] = {
+    config.staticProjectsInPerimeters.getOrElse(servicePerimeterName, Seq.empty)
+  }
+
+  /**
+    * Takes the the name of a Service Perimeter as the only parameter.  Since multiple Billing Projects can specify the
+    * same Service Perimeter, we will:
+    * 1. Load all the Billing Projects that also use this servicePerimeterName
+    * 2. Load all the Workspaces in all of those Billing Projects
+    * 3. Collect all of the GoogleProjectNumbers from those Workspaces
+    * 4. Post that list to Google to overwrite the Service Perimeter's list of included Google Projects
+    * 5. Poll until Google Operation to update the Service Perimeter gets to some terminal state
+    * Throw exceptions if any of this goes awry
+    *
+    * @param servicePerimeterName
+    * @return Future[Unit] indicating whether we succeeded to update the Service Perimeter
+    */
+  private def overwriteGoogleProjectsInPerimeter(servicePerimeterName: ServicePerimeterName): Future[Unit] = {
+    collectWorkspacesInPerimeter(servicePerimeterName).map { workspacesInPerimeter =>
+      val projectNumbers = workspacesInPerimeter.flatMap(_.googleProjectNumber) ++ loadStaticProjectsForPerimeter(servicePerimeterName)
+      val projectNumberStrings = projectNumbers.map(_.value)
```

Comment:
should dedup `projectNumberStrings`, with v1 workspaces there could be duplicates

Reply:
changed the signature for `AccessContextManagerDAO.overwriteProjectsInServicePerimeter` to require a `Set` of project numbers instead of a `Seq`